pydantic/fields.py

```diff
@@ -416,6 +420,12 @@ def _type_analysis(self) -> None:  # noqa: C901 (ignore complexity)
             self.key_field = self._create_sub_type(self.type_.__args__[0], 'key_' + self.name, for_keys=True)
             self.type_ = self.type_.__args__[1]
             self.shape = SHAPE_MAPPING
+        # Equality check as almost everything inherit form Iterable, including str
+        # check for typing.Iterable and abc.Iterable, as it could receive one even when declared with the other
+        elif origin == Iterable or origin == abc.Iterable:
```

Comment:
```suggestion
        elif origin in {Iterable, CollectionsIterable}:
```

or is this slower in cython?

Reply:
No idea but this looks cleaner. Let's try it.